File: frontend/src/api/video.ts
import { apiClient } from "./client";

export interface VideoTask {
  id: number;
  url: string;
  final_url?: string;
  platform?: string;
  title?: string;
  author?: string;
  description?: string;
  thumbnail_url?: string;
  duration?: string;
  published_at?: string;
  view_count?: string;
  direct_video_url?: string;
  download_requested: boolean;
  download_status: "not_requested" | "pending" | "completed" | "blocked" | "failed";
  download_path?: string;
  download_url?: string;
  status: "pending" | "fetching" | "completed" | "blocked" | "failed";
  error_message?: string;
  created_at: string;
  updated_at: string;
}

export function parseVideoUrl(url: string, download = false) {
  return apiClient.post<VideoTask>("/video/parse", { url, download });
}

export function fetchVideoTasks() {
  return apiClient.get<VideoTask[]>("/video/tasks");
}

export function fetchVideoTask(taskId: number) {
  return apiClient.get<VideoTask>(`/video/tasks/${taskId}`);
}

export function videoDownloadUrl(task: VideoTask) {
  if (!task.download_url) {
    return undefined;
  }
  const baseURL = apiClient.defaults.baseURL ?? "";
  return task.download_url.startsWith("http") ? task.download_url : `${baseURL}${task.download_url.replace(/^\/api/, "")}`;
}

File: backend/app/schemas/video.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, HttpUrl


class VideoParseRequest(BaseModel):
    url: HttpUrl
    download: bool = False


class VideoTaskRead(BaseModel):
    id: int
    url: str
    final_url: str | None = None
    platform: str | None = None
    title: str | None = None
    author: str | None = None
    description: str | None = None
    thumbnail_url: str | None = None
    duration: str | None = None
    published_at: str | None = None
    view_count: str | None = None
    direct_video_url: str | None = None
    download_requested: bool
    download_status: str
    download_path: str | None = None
    status: str
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
